pkg/storage/backend-storage/backend-storage.go

```diff
@@ -71,24 +74,127 @@ func IsBackendStorageNeededForVM(vm *corev1.VirtualMachine) bool {
 	return HasPersistentTPMDevice(&vm.Spec.Template.Spec)
 }
 
-func CreateIfNeeded(vmi *corev1.VirtualMachineInstance, clusterConfig *virtconfig.ClusterConfig, client kubecli.KubevirtClient) error {
+type BackendStorage struct {
+	client          kubecli.KubevirtClient
+	clusterConfig   *virtconfig.ClusterConfig
+	scStore         cache.Store
+	accessModeCache map[string]v1.PersistentVolumeAccessMode
+	cacheLock       sync.Mutex
+}
+
+func NewBackendStorage(client kubecli.KubevirtClient, clusterConfig *virtconfig.ClusterConfig, scStore cache.Store) *BackendStorage {
+	return &BackendStorage{
+		client:          client,
+		clusterConfig:   clusterConfig,
+		scStore:         scStore,
+		accessModeCache: map[string]v1.PersistentVolumeAccessMode{},
+	}
+}
+
+func (bs *BackendStorage) getStorageClass() (string, error) {
+	storageClass := bs.clusterConfig.GetVMStateStorageClass()
+	if storageClass != "" {
+		return storageClass, nil
+	}
+
+	for _, obj := range bs.scStore.List() {
+		sc := obj.(*storagev1.StorageClass)
+		if sc.Annotations["storageclass.kubernetes.io/is-default-class"] == "true" {
+			return sc.Name, nil
+		}
+	}
+
+	return "", fmt.Errorf("no default storage class found")
+}
+
+func (bs *BackendStorage) getAccessMode(storageClass string, mode v1.PersistentVolumeMode) v1.PersistentVolumeAccessMode {
+	bs.cacheLock.Lock()
+	defer bs.cacheLock.Unlock()
+	if accessMode, exists := bs.accessModeCache[storageClass]; exists {
+		return accessMode
+	}
+
+	// Storage profiles are guaranteed to have the same name as their storage class
+	storageProfile, err := bs.client.CdiClient().CdiV1beta1().StorageProfiles().Get(context.Background(), storageClass, metav1.GetOptions{})
+	if errors.IsNotFound(err) {
+		log.Log.Infof("no storage profile found for %s, defaulting to RWX", storageClass)
+		return v1.ReadWriteMany
+	}
+	if err != nil {
+		// CDI not present most likely
+		log.Log.Reason(err).Info("couldn't access storage profiles, defaulting to RWX")
+		return v1.ReadWriteMany
+	}
+
+	bs.accessModeCache[storageClass] = v1.ReadWriteMany
+	if storageProfile.Status.ClaimPropertySets == nil || len(storageProfile.Status.ClaimPropertySets) == 0 {
+		log.Log.Infof("no ClaimPropertySets in storage profile %s, defaulting to RWX", storageProfile.Name)
+		return v1.ReadWriteMany
```

Comment:
This does not seem like a reasonable default to me, if this is empty I think either

1) CDI is still determining what works, so requeue reconcile in a bit may work
2) CDI tried and provisioner is broken/can't allocate PVCs

Reply:
> All that being said, I still think it is unreasonable to assume the default StorageClass is ReadWriteMany Filesystem. Would go with ReadWriteOnce

Defaulting to RWX is actually current behavior. Changing to a RWO default would break existing use-cases. RWO makes the VM un-migratable, which is a big deal. (Obviously we can't assume all clusters have storage profiles).
As documented, users who wish to use persistent EFI/TPM should configure a storage class in the CR that's capable of FS in RWX. If they don't, they understand the risk that their VMs may not start.

Reply:
To me default RWX makes sense if the user configured the storageclass on the kubevirt CR.  I am suggesting that it would be very risky to assume that the default storage class for the cluster supports RWX filesystem

Reply:
So I guess what I'm suggesting is:  in the case where StorageProfile is not helping, let's go with RWX if the user explicitly set the StorageClass, RWO otherwise.

Before committing to RWO one thing you could do is list all PVs for that StorageClass and see if there's a RWX one?

Reply:
> So I guess what I'm suggesting is: in the case where StorageProfile is not helping, let's go with RWX if the user explicitly set the StorageClass, RWO otherwise.

Pushed, PTAL!

> Before committing to RWO one thing you could do is list all PVs for that StorageClass and see if there's a RWX one?

Interesting idea, but we need this to be deterministic...

Reply:
Why does it have to be deterministic?

Reply:
> Plus we'd probably want to use a pvInformer, in which case the above might not even be true!

Not sure what you mean by that but this is an optimization we can hold off on for now

Reply:
A pvInformer might not yet know about a RWX pv that was just created...
Anyway, yeah, let's keep that as an idea to explore in the future.